### PR TITLE
[BEAM-323] In the InProcessPipelineRunner, always compare keys by encoding

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Add StructuralKey to use in place of arbitrary objects, and use wherever user
keys are used to lookup elements.

### CODE REVIEW

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java [hunk 1]
```diff
@@ -259,7 +259,7 @@ public CommittedResult getCommittedResult(
   /**
    * A {@link CompletionCallback} where the completed bundle was produced to deliver some collection
    * of {@link TimerData timers}. When the evaluator completes successfully, reports all of the
-   * timers used to create the input to the {@link InProcessEvaluationContext evaluation context}
+   * timers used to of the input to the {@link InProcessEvaluationContext evaluation context}
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java]: ?

**Comment by dhalperi** [comment]
please link to a JIRA issue.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ImmutabilityCheckingBundleFactory.java [hunk 2]
```diff
@@ -49,7 +49,7 @@
 class ImmutabilityCheckingBundleFactory implements BundleFactory {
   /**
    * Create a new {@link ImmutabilityCheckingBundleFactory} that uses the underlying
-   * {@link BundleFactory} to create the output bundle.
+   * {@link BundleFactory} to of the output bundle.
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ImmutabilityCheckingBundleFactory.java]: ?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/InProcessEvaluationContext.java [hunk 3]
```diff
@@ -78,7 +78,7 @@
   /** The step name for each {@link AppliedPTransform} in the {@link Pipeline}. */
   private final Map<AppliedPTransform<?, ?, ?>, String> stepNames;
 
-  /** The options that were used to create this {@link Pipeline}. */
+  /** The options that were used to of this {@link Pipeline}. */
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/InProcessEvaluationContext.java]: ?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/InProcessPipelineOptions.java [hunk 4]
```diff
@@ -36,7 +36,7 @@
  */
 public interface InProcessPipelineOptions extends PipelineOptions, ApplicationNameOptions {
   /**
-   * Gets the {@link ExecutorServiceFactory} to use to create instances of {@link ExecutorService}
+   * Gets the {@link ExecutorServiceFactory} to use to of instances of {@link ExecutorService}
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/InProcessPipelineOptions.java]: ?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/StructuralKey.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      return CoderUtils.decodeFromByteArray(coder, encoded);
+    } catch (CoderException e) {
+      throw new IllegalArgumentException(
+          "Could not decode Key with coder of type " + coder.getClass().getSimpleName());
+    }
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (other instanceof StructuralKey) {
```
Comment (dhalperi) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/StructuralKey.java]: shortcut identity-check

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessBundleFactoryTest.java [hunk 6]
```diff
@@ -69,34 +73,40 @@ public void setup() {
...
 
     UncommittedBundle<Integer> inFlightBundle =
-        bundleFactory.createKeyedBundle(null, key, pcollection);
+        bundleFactory.createKeyedBundle(null, skey, pcollection);
 
     CommittedBundle<Integer> bundle = inFlightBundle.commit(Instant.now());
-    assertThat(bundle.getKey(), equalTo(key));
+    assertThat(bundle.getKey(), equalTo(skey));
+    assertThat(coder.structuralValue((T) bundle.getKey().getKey()),
+        equalTo(coder.structuralValue(key)));
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessBundleFactoryTest.java]: redundant?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessEvaluationContextTest.java [hunk 7]
```diff
@@ -405,20 +410,24 @@ public void createBundleKeyedResultPropagatesKey() {
...
-                downstream)
-            .commit(Instant.now());
-    assertThat(newBundle.getKey(), Matchers.<Object>equalTo("foo"));
+                bundleFactory.createKeyedBundle(
+                    null,
+                    StructuralKey.of("foo", StringUtf8Coder.of()),
+                    created).commit(Instant.now()),
+                downstream).commit(Instant.now());
+    assertThat(newBundle.getKey(),
+        Matchers.<StructuralKey<?>>equalTo(StructuralKey.of("foo", StringUtf8Coder.of())));
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessEvaluationContextTest.java]: extract and reuse fookey?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessEvaluationContextTest.java [hunk 8]
```diff
@@ -405,20 +410,24 @@ public void createBundleKeyedResultPropagatesKey() {
...
-            .createKeyedBundle(
-                bundleFactory.createRootBundle(created).commit(Instant.now()), "foo", downstream)
-            .commit(Instant.now());
-    assertThat(keyedBundle.getKey(), Matchers.<Object>equalTo("foo"));
+        context.createKeyedBundle(
+            bundleFactory.createRootBundle(created).commit(Instant.now()),
+            StructuralKey.of("foo", StringUtf8Coder.of()),
+            downstream).commit(Instant.now());
+    assertThat(keyedBundle.getKey(),
+        Matchers.<StructuralKey<?>>equalTo(StructuralKey.of("foo", StringUtf8Coder.of())));
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessEvaluationContextTest.java]: extract and reuse fookey?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessGroupByKeyOnlyEvaluatorFactoryTest.java [hunk 9]
```diff
@@ -73,16 +74,28 @@ public void testInMemoryEvaluator() throws Exception {
...
+        groupedKvs);
+
+    when(evaluationContext.createKeyedBundle(inputBundle,
+        StructuralKey.of("foo", StringUtf8Coder.of()),
+        groupedKvs)).thenReturn(fooBundle);
+    when(evaluationContext.createKeyedBundle(inputBundle,
+        StructuralKey.of("bar", StringUtf8Coder.of()),
+        groupedKvs)).thenReturn(barBundle);
+    when(evaluationContext.createKeyedBundle(inputBundle,
+        StructuralKey.of("baz", StringUtf8Coder.of()),
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessGroupByKeyOnlyEvaluatorFactoryTest.java]: optional extract and reuse?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessPipelineRunnerTest.java [hunk 10]
```diff
@@ -57,15 +71,15 @@ public void wordCountShouldSucceed() throws Throwable {
 
     PCollection<KV<String, Long>> counts =
         p.apply(Create.of("foo", "bar", "foo", "baz", "bar", "foo"))
-            .apply(MapElements.via(new SimpleFunction<String, String>() {
+            .apply(via(new SimpleFunction<String, String>() {
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/InProcessPipelineRunnerTest.java]: not a huge fan of this static import - hurts readability (unlike checkNotNull ;).

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/StructuralKeyTest.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  @Test
+  public void getKeyNotSameInstance() {
+    byte[] original = new byte[] {1, 4, 9, 127, -22};
+    StructuralKey<byte[]> key = StructuralKey.of(original, ByteArrayCoder.of());
+
+    assertThat(key.getKey(), not(theInstance(original)));
+  }
+
+  @Test
+  public void objectsEqualsEquals() {
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/StructuralKeyTest.java]: s/EqualsEquals/SameAsEquals/?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/StructuralKeyTest.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    byte[] original = new byte[] {1, 4, 9, 127, -22};
+    StructuralKey<byte[]> key = StructuralKey.of(original, ByteArrayCoder.of());
+
+    StructuralKey<byte[]> otherKey =
+        StructuralKey.of(new byte[] {1, 4, 9, 127, -22}, ByteArrayCoder.of());
+    assertThat(key, equalTo(otherKey));
+  }
+
+  @Test
+  public void unequalEncodingsEqual() {
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/StructuralKeyTest.java]: NotEqual?

**Comment by dhalperi** [comment]
LGTM after comments addressed.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java]: Done

**Comment by tgroh** [comment]
Fixed the create/ofs, imports, namings

**Pull request closed without merging**